packages/upload/src/index.ts

```diff
@@ -0,0 +1,182 @@
+};
+
+export interface FilesUploader {
+  files: Accessor<UploadFile[]>;
+  selectFiles: (callback: (file: UploadFile[]) => void | Promise<void>) => void;
+  /**
+   * `change` event handler for input element. Handles `drop` event as well
+   */
+  handleFilesInput: (
+    event: (Event | DragEvent) & {
+      currentTarget: HTMLInputElement;
+      target: Element;
+    }
+  ) => void | Promise<void>;
+  removeFile: (fileName: string) => void;
+  clearFiles: () => void;
+}
+
+export interface FileUploader extends Omit<FilesUploader, "files" | "selectFiles"> {
+  files: Accessor<UploadFile>;
+  selectFiles: (callback: (file: UploadFile) => void | Promise<void>) => void;
+}
+
+/**
+ * Primitive to make uploading files easier.
+ *
+ * @returns `file (files)`
+ * @returns `selectFile (selectFiles)` - Open file picker and set files
+ * @returns `handleFileInput (handleFilesInput)` - Event handler for `<input type="file" />`
+ * @returns `removeFile`
+ * @returns `clearFiles`
+ *
+ * @example
+ * ```ts
+ * // multiple files
+ * const {files, selectFiles} = createFileUploader({ multiple: true, accept: "image/*" });
+ * selectFiles(files => {
+ *   console.log(files);
+ * });
+ *
+ * // single file
+ * const {file, selectFile} = createFileUploader();
+ * selectFile(({ source, name, size, file }) => {
+ *   console.log({ source, name, size, file });
+ * });
+ * ```
+ */
+
+function createFileUploader(): FileUploader;
+function createFileUploader(options?: CreateFileUploaderOptions): FilesUploader;
+function createFileUploader(options?: CreateFileUploaderOptions): any {
+  const [files, setFiles] = createSignal<UploadFile | UploadFile[]>();
+
+  let userCallback: FileSelectCallback;
+
+  async function onChange(this: HTMLInputElement, ev: Event) {
+    const parsedFiles = [];
+    const target = this;
+
+    for (const index in target.files) {
+      const fileIndex = +index;
+      if (isNaN(+fileIndex)) {
+        continue;
+      }
+
+      const file = target.files[fileIndex];
+      const parsedFile = transformFile(file);
+
+      parsedFiles.push(parsedFile);
+    }
+
+    target.removeEventListener("change", onChange);
+
+    target.remove();
+
+    if (target.multiple) {
+      setFiles(parsedFiles);
+      return userCallback(parsedFiles);
+    }
+
+    setFiles(parsedFiles[0]);
+    return userCallback(parsedFiles[0]);
+  }
+
+  const selectFiles = (callback: FileSelectCallback) => {
+    userCallback = callback;
+
+    if (callback instanceof Promise) {
+      callback.then().catch(e => console.error(e));
+    }
+
+    const inputEL = createInputComponent(options || {});
+
+    inputEL.addEventListener("change", onChange);
+    inputEL.click();
+  };
+
+  const handleFilesInput = async (
+    event: (Event | DragEvent) & {
+      currentTarget: HTMLInputElement;
+      target: Element;
+    }
+  ) => {
+    const filesArr: File[] = Array.from(
+      event?.currentTarget?.files || (event as DragEvent)?.dataTransfer?.files || []
```

Comment:
Seems like I can use `change` event handler when I drag'n'drop files, so I decided to handle drop event in one handler fn. Would it be better if made a separate `createDropzone` utility as @davedbase suggested?

Reply:
Sorry for the delay! It's been a busy start to the week! I think a separate dropzone would be nice compliment or addition to this package.